File: sdks/official/fraiseql-csharp/tests/FraiseQL.Tests/ClientTests.cs
using System.Net;
using System.Text;
using Xunit;

namespace FraiseQL.Tests;

/// <summary>
/// Tests for <see cref="FraiseQLClient"/> using a stub <see cref="HttpMessageHandler"/>
/// so no real network connection is required.
/// </summary>
public sealed class ClientTests
{
    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static FraiseQLClient BuildClient(HttpStatusCode status, string body)
    {
        var handler = new StubHandler(status, body);
        var http = new HttpClient(handler) { BaseAddress = null };
        var options = new FraiseQLClientOptions
        {
            Url = "http://localhost/graphql",
            HttpClient = http,
        };
        return new FraiseQLClient(options);
    }

    private static string DataJson(string payload)
        => $"{{\"data\":{payload}}}";

    private static string ErrorJson(string message)
        => $"{{\"errors\":[{{\"message\":\"{message}\"}}]}}";

    private static string NullErrorsJson(string payload)
        => $"{{\"data\":{payload},\"errors\":null}}";

    // ------------------------------------------------------------------
    // QueryAsync — success
    // ------------------------------------------------------------------

    [Fact]
    public async Task QueryAsync_ReturnsData_OnSuccess()
    {
        using var client = BuildClient(HttpStatusCode.OK, DataJson("{\"id\":42}"));

        var result = await client.QueryAsync<UserResult>("{ user { id } }");

        Assert.NotNull(result);
        Assert.Equal(42, result.Id);
    }

    [Fact]
    public async Task QueryAsync_ReturnsDefaultT_WhenDataIsNull()
    {
        using var client = BuildClient(HttpStatusCode.OK, "{\"data\":null}");

        var result = await client.QueryAsync<UserResult>("{ user { id } }");

        Assert.Null(result);
    }

    // ------------------------------------------------------------------
    // QueryAsync — null errors field treated as success
    // ------------------------------------------------------------------

    [Fact]
    public async Task QueryAsync_DoesNotThrow_WhenErrorsFieldIsNull()
    {
        using var client = BuildClient(HttpStatusCode.OK, NullErrorsJson("{\"id\":1}"));

        // Must NOT throw — null errors field is not an error
        var result = await client.QueryAsync<UserResult>("{ user { id } }");

        Assert.NotNull(result);
        Assert.Equal(1, result.Id);
    }

    [Fact]
    public async Task QueryAsync_DoesNotThrow_WhenErrorsArrayIsEmpty()
    {
        using var client = BuildClient(HttpStatusCode.OK, "{\"data\":{\"id\":7},\"errors\":[]}");

        var result = await client.QueryAsync<UserResult>("{ user { id } }");

        Assert.Equal(7, result.Id);
    }

    // ------------------------------------------------------------------
    // QueryAsync — GraphQL errors
    // ------------------------------------------------------------------

    [Fact]
    public async Task QueryAsync_ThrowsGraphQLException_WhenErrorsPresent()
    {
        using var client = BuildClient(HttpStatusCode.OK, ErrorJson("field not found"));

        var ex = await Assert.ThrowsAsync<GraphQLException>(
            () => client.QueryAsync<object>("{ bad }"));

        Assert.Single(ex.Errors);
        Assert.Equal("field not found", ex.Errors[0].Message);
    }

    [Fact]
    public async Task QueryAsync_ExceptionMessage_EqualsFirstErrorMessage()
    {
        using var client = BuildClient(HttpStatusCode.OK,
            "{\"errors\":[{\"message\":\"first\"},{\"message\":\"second\"}]}");

        var ex = await Assert.ThrowsAsync<GraphQLException>(
            () => client.QueryAsync<object>("{ bad }"));

        Assert.Equal("first", ex.Message);
        Assert.Equal(2, ex.Errors.Count);
    }

    // ------------------------------------------------------------------
    // QueryAsync — HTTP error status codes
    // ------------------------------------------------------------------

    [Fact]
    public async Task QueryAsync_ThrowsAuthenticationException_On401()
    {
        using var client = BuildClient(HttpStatusCode.Unauthorized, "Unauthorized");

        var ex = await Assert.ThrowsAsync<AuthenticationException>(
            () => client.QueryAsync<object>("{ x }"));

        Assert.Equal(401, ex.StatusCode);
    }

    [Fact]
    public async Task QueryAsync_ThrowsAuthenticationException_On403()
    {
        using var client = BuildClient(HttpStatusCode.Forbidden, "Forbidden");

        var ex = await Assert.ThrowsAsync<AuthenticationException>(
            () => client.QueryAsync<object>("{ x }"));

        Assert.Equal(403, ex.StatusCode);
    }

    [Fact]
    public async Task QueryAsync_ThrowsRateLimitException_On429()
    {
        using var client = BuildClient((HttpStatusCode)429, "Too Many Requests");

        await Assert.ThrowsAsync<RateLimitException>(
            () => client.QueryAsync<object>("{ x }"));
    }

    // ------------------------------------------------------------------
    // Synchronous wrappers
    // ------------------------------------------------------------------

    [Fact]
    public void Query_ReturnsData_Synchronously()
    {
        using var client = BuildClient(HttpStatusCode.OK, DataJson("{\"id\":99}"));

        var result = client.Query<UserResult>("{ user { id } }");

        Assert.Equal(99, result.Id);
    }

    [Fact]
    public void Mutate_ReturnsData_Synchronously()
    {
        using var client = BuildClient(HttpStatusCode.OK, DataJson("{\"id\":5}"));

        var result = client.Mutate<UserResult>("mutation { createUser { id } }");

        Assert.Equal(5, result.Id);
    }

    // ------------------------------------------------------------------
    // Authorization header
    // ------------------------------------------------------------------

    [Fact]
    public async Task QueryAsync_SendsStaticAuthorizationHeader()
    {
        var handler = new CapturingHandler(HttpStatusCode.OK, DataJson("null"));
        var http = new HttpClient(handler);
        using var client = new FraiseQLClient(new FraiseQLClientOptions
        {
            Url = "http://localhost/graphql",
            Authorization = "Bearer my-token",
            HttpClient = http,
        });

        await client.QueryAsync<object>("{ x }");

        Assert.True(handler.LastRequest?.Headers.Contains("Authorization"));
        Assert.Equal("Bearer my-token",
            handler.LastRequest?.Headers.GetValues("Authorization").First());
    }

    [Fact]
    public async Task QueryAsync_SendsDynamicAuthorizationHeader_WhenFactorySet()
    {
        var handler = new CapturingHandler(HttpStatusCode.OK, DataJson("null"));
        var http = new HttpClient(handler);
        using var client = new FraiseQLClient(new FraiseQLClientOptions
        {
            Url = "http://localhost/graphql",
            AuthorizationFactory = () => Task.FromResult("Bearer dynamic"),
            HttpClient = http,
        });

        await client.QueryAsync<object>("{ x }");

        Assert.Equal("Bearer dynamic",
            handler.LastRequest?.Headers.GetValues("Authorization").First());
    }

    // ------------------------------------------------------------------
    // Helpers — stub types
    // ------------------------------------------------------------------

    private sealed record UserResult(int Id);

    private sealed class StubHandler : HttpMessageHandler
    {
        private readonly HttpStatusCode _status;
        private readonly string _body;

        public StubHandler(HttpStatusCode status, string body)
        {
            _status = status;
            _body = body;
        }

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            var response = new HttpResponseMessage(_status)
            {
                Content = new StringContent(_body, Encoding.UTF8, "application/json")
            };
            return Task.FromResult(response);
        }
    }

    private sealed class CapturingHandler : HttpMessageHandler
    {
        private readonly HttpStatusCode _status;
        private readonly string _body;

        public HttpRequestMessage? LastRequest { get; private set; }

        public CapturingHandler(HttpStatusCode status, string body)
        {
            _status = status;
            _body = body;
        }

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            LastRequest = request;
            var response = new HttpResponseMessage(_status)
            {
                Content = new StringContent(_body, Encoding.UTF8, "application/json")
            };
            return Task.FromResult(response);
        }
    }
}

File: sdks/official/fraiseql-fsharp/tests/FraiseQL.Tests/ClientTests.fs
module FraiseQL.Tests.ClientTests

open System
open System.Net
open System.Net.Http
open System.Text
open System.Threading
open System.Threading.Tasks
open Xunit
open FraiseQL

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

type UserResult() =
    member val id: int = 0 with get, set

/// Stub HttpMessageHandler that always returns a fixed response.
type StubHandler(status: HttpStatusCode, body: string) =
    inherit HttpMessageHandler()
    override _.SendAsync(_, _) =
        let response = new HttpResponseMessage(status,
                           Content = new StringContent(body, Encoding.UTF8, "application/json"))
        Task.FromResult response

/// HttpMessageHandler that captures the last request sent.
type CapturingHandler(status: HttpStatusCode, body: string) =
    inherit HttpMessageHandler()
    let mutable lastRequest: HttpRequestMessage option = None
    member _.LastRequest = lastRequest
    override _.SendAsync(request, _) =
        lastRequest <- Some request
        let response = new HttpResponseMessage(status,
                           Content = new StringContent(body, Encoding.UTF8, "application/json"))
        Task.FromResult response

let private dataJson payload = $"""{"{"}"data":{payload}{"}"}"""
let private errorJson msg   = $"""{"{"}"errors":[{"{"}"message":"{msg}"{"}"}]{"}"}"""

let private buildClient status body =
    let handler = new StubHandler(status, body)
    let http    = new HttpClient(handler)
    let options = { FraiseQLClientOptions.defaultFor "http://localhost/graphql"
                    with HttpClient = Some http }
    new FraiseQLHttpClient(options)

// ---------------------------------------------------------------------------
// QueryAsync — success
// ---------------------------------------------------------------------------

[<Fact>]
let ``ExecuteAsync returns data on success`` () =
    let json = dataJson """{"id":42}"""
    use client = buildClient HttpStatusCode.OK json
    let result = client.ExecuteAsync<UserResult>("{ user { id } }").GetAwaiter().GetResult()
    Assert.Equal(42, result.id)


[<Fact>]
let ``AsyncQuery returns data on success`` () =
    async {
        use client = buildClient HttpStatusCode.OK (dataJson """{"id":42}""")
        let! result = client.AsyncQuery<UserResult>("{ user { id } }")
        Assert.Equal(42, result.id)
    } |> Async.RunSynchronously

[<Fact>]
let ``AsyncQuery returns default when data is null`` () =
    async {
        use client = buildClient HttpStatusCode.OK """{"data":null}"""
        let! result = client.AsyncQuery<UserResult option>("{ user { id } }")
        Assert.Equal(None, result)
    } |> Async.RunSynchronously

// ---------------------------------------------------------------------------
// QueryAsync — null errors field treated as success (CRITICAL)
// ---------------------------------------------------------------------------

[<Fact>]
let ``AsyncQuery does not throw when errors field is null`` () =
    async {
        use client = buildClient HttpStatusCode.OK """{"data":{"id":1},"errors":null}"""
        // Must NOT throw — null errors is not an error
        let! result = client.AsyncQuery<UserResult>("{ user { id } }")
        Assert.Equal(1, result.id)
    } |> Async.RunSynchronously

[<Fact>]
let ``AsyncQuery does not throw when errors array is empty`` () =
    async {
        use client = buildClient HttpStatusCode.OK """{"data":{"id":7},"errors":[]}"""
        let! result = client.AsyncQuery<UserResult>("{ user { id } }")
        Assert.Equal(7, result.id)
    } |> Async.RunSynchronously

// ---------------------------------------------------------------------------
// QueryAsync — GraphQL errors
// ---------------------------------------------------------------------------

[<Fact>]
let ``AsyncQuery throws GraphQLException when errors present`` () =
    async {
        use client = buildClient HttpStatusCode.OK (errorJson "field not found")
        let! ex = Assert.ThrowsAsync<GraphQLException>(
                      fun () -> client.ExecuteAsync<obj>("{ bad }"))
                  |> Async.AwaitTask
        Assert.Single(ex.Errors) |> ignore
        Assert.Equal("field not found", ex.Errors.[0].Message)
    } |> Async.RunSynchronously

// ---------------------------------------------------------------------------
// QueryAsync — HTTP errors
// ---------------------------------------------------------------------------

[<Fact>]
let ``AsyncQuery throws AuthenticationException on 401`` () =
    async {
        use client = buildClient HttpStatusCode.Unauthorized "Unauthorized"
        let! ex = Assert.ThrowsAsync<AuthenticationException>(
                      fun () -> client.ExecuteAsync<obj>("{ x }"))
                  |> Async.AwaitTask
        Assert.Equal(401, ex.StatusCode)
    } |> Async.RunSynchronously

[<Fact>]
let ``AsyncQuery throws AuthenticationException on 403`` () =
    async {
        use client = buildClient HttpStatusCode.Forbidden "Forbidden"
        let! ex = Assert.ThrowsAsync<AuthenticationException>(
                      fun () -> client.ExecuteAsync<obj>("{ x }"))
                  |> Async.AwaitTask
        Assert.Equal(403, ex.StatusCode)
    } |> Async.RunSynchronously

[<Fact>]
let ``AsyncQuery throws RateLimitException on 429`` () =
    async {
        use client = buildClient (enum<HttpStatusCode> 429) "Too Many Requests"
        do! Assert.ThrowsAsync<RateLimitException>(
                fun () -> client.ExecuteAsync<obj>("{ x }"))
            |> Async.AwaitTask
            |> Async.Ignore
    } |> Async.RunSynchronously

// ---------------------------------------------------------------------------
// Authorization headers
// ---------------------------------------------------------------------------

[<Fact>]
let ``ExecuteAsync sends static Authorization header`` () =
    async {
        let handler = new CapturingHandler(HttpStatusCode.OK, dataJson "null")
        let http    = new HttpClient(handler)
        use client  = new FraiseQLHttpClient(
                          { FraiseQLClientOptions.defaultFor "http://localhost/graphql"
                            with Authorization = Some "Bearer my-token"
                                 HttpClient    = Some http })
        let! _ = client.ExecuteAsync<obj>("{ x }") |> Async.AwaitTask
        let req = handler.LastRequest.Value
        Assert.True(req.Headers.Contains "Authorization")
        Assert.Equal("Bearer my-token",
            req.Headers.GetValues("Authorization") |> Seq.head)
    } |> Async.RunSynchronously
